java_src/foam/nanos/auth/AuthOAO.java

```diff
@@ -0,0 +1,65 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.oao.ProxyOAO;
+
+import javax.security.auth.AuthPermission;
+import java.util.Map;
+
+public class AuthOAO
+  extends ProxyOAO
+{
+  protected final String rootPermission_;
+
+  public AuthOAO() {
+    this.rootPermission_ = null;
+  }
+
+  public AuthOAO(String rootPermission_) {
+    this.rootPermission_ = rootPermission_;
+  }
+
+  protected String getRootPermission_(X x) throws IllegalAccessException {
+    return rootPermission_ == null ? getDelegate().get(x).getClassInfo().getId() : rootPermission_;
+  }
+
+  @Override
+  public FObject get(X x) throws IllegalAccessException {
+    AuthService authService = (AuthService) x.get("authService");
+    java.security.Permission permission = new AuthPermission(getRootPermission_(x));
+    if ( ! authService.check(x, permission) ) {
+      throw new IllegalAccessException("Invalid permissions");
+    }
+    return super.get(x);
+  }
+
+  @Override
+  public void setProperty(X x, String name, Object value) throws IllegalAccessException {
+    AuthService authService = (AuthService) x.get("authService");
+    java.security.Permission permission = new AuthPermission(getRootPermission_(x) + "." + name);
```

Comment:
extract the next five lines to a setProperty_() method so you can call it from both setProperty() and setProperties() to avoid duplicating the code.